File: java/src/main/java/com/haibo/yan/design/amazon/Amazon.java
/*
 *
 *  * Copyright [2017] [Haibo(Tristan) Yan]
 *  *
 *  * Licensed under the Apache License, Version 2.0 (the "License");
 *  * you may not use this file except in compliance with the License.
 *  * You may obtain a copy of the License at
 *  *
 *  *     http://www.apache.org/licenses/LICENSE-2.0
 *  *
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS,
 *  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  * See the License for the specific language governing permissions and
 *  * limitations under the License.
 *
 */

// see https://github.com/billyean/algo_system_design/blob/master/amazon_1.png
// and https://github.com/billyean/algo_system_design/blob/master/amazon_2.png
package com.haibo.yan.design.amazon;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Amazon {
    public List<String> retrieveMostFrequentlyUsedWords(String literatureText,
                                                        List<String> wordsToExclude) {
        // WRITE YOUR CODE HERE
        Set<String> setToExcluded = new HashSet<>();
        for (String excluded : wordsToExclude) {
            if (excluded != null) {
                setToExcluded.add(excluded.trim().toLowerCase());
            }
        }

        Map<String, Integer> wordsMap = new HashMap<>();
        String[] words = literatureText.split("[[^a-z]&&[^A-Z]]");

        for (String word : words) {
            String low = word.toLowerCase();
            System.out.println(low);
            if (!setToExcluded.contains(low)) {
                wordsMap.put(low, wordsMap.getOrDefault(low, 0) + 1);
            }
        }
        List<String> mostFrequent = new ArrayList<>();
        int toppest = 0;
        for (int frequency : wordsMap.values()) {
            toppest = Math.max(toppest, frequency);
        }
        for (Map.Entry<String, Integer> entry : wordsMap.entrySet()) {
            if (entry.getValue() == toppest) {
                mostFrequent.add(entry.getKey());
            }
        }
        Collections.sort(mostFrequent);
        return mostFrequent;
    }

    public List<String> reorderLines(int logFileSize, List<String> logLines) {
        // WRITE YOUR CODE HERE
        List<String> words = new ArrayList<>();
        if (logFileSize == 0 || logLines.size() == 0) {
            return words;
        }

        List<String> integers = new ArrayList<>();

        for (String fline : logLines) {
            String line = fline.trim();
            int sp1 = line.indexOf(' ');
            int sp2 = line.indexOf(' ', sp1 + 1);
            if (line.substring(sp1 + 1, sp2).trim().matches("\\d+")) {
                integers.add(line);
            } else {
                words.add(line);
            }
        }

        Collections.sort(words, (s1, s2) -> {
            String r1 = s1.substring(s1.indexOf(' ')).trim();
            String r2 = s2.substring(s2.indexOf(' ')).trim();

            if (r1.equals(r2)) {
                String identifier1 = s1.substring(0, s1.indexOf(' '));
                String identifier2 = s2.substring(0, s2.indexOf(' '));

                Pattern p = Pattern.compile("([a-z]+)(\\d+)");
                Matcher m1 = p.matcher(identifier1);
                Matcher m2 = p.matcher(identifier2);

                if (m1.find() && m2.find()) {
                    int i1 = Integer.parseInt(m1.group(2));
                    int i2 = Integer.parseInt(m2.group(2));

                    String is1 = m1.group(1);
                    String is2 = m2.group(1);

                    if (i1 == i2) {
                        return is1.compareTo(is2);
                    } else {
                        return Integer.compare(i1, i2);
                    }
                }

                return 0;
            } else {
                return r1.compareTo(r2);
            }
        });

        words.addAll(integers);
        return words;
    }
}

File: c++/algorithm/str/Amazon.cpp
//
// Created by Haibo Yan on 3/31/18.
//

// see https://github.com/billyean/algo_system_design/blob/master/amazon_1.png
// and https://github.com/billyean/algo_system_design/blob/master/amazon_2.png
#include "Amazon.h"

#include <map>
#include <regex>
#include <set>
#include <iostream>
#include <algorithm>

vector<string> Amazon::retrieveMostFrequentlyUsedWords(string literatureText, vector<string> wordsToExclude) {
    set<string> exclude;

    for (auto iw = wordsToExclude.begin(); iw != wordsToExclude.end(); iw++) {
        string w = (*iw);
        transform(w.begin(), w.end(), w.begin(), ::tolower);
        exclude.insert(w);x
    }

    map<string, int> words;
    regex r("([A-Za-z]+)");

    sregex_iterator next(literatureText.begin(), literatureText.end(), r), end;

    while (next != end) {
        string w = next->str();
        transform(w.begin(), w.end(), w.begin(), ::tolower);
        if (find(exclude.begin(), exclude.end(), w) == exclude.end()) {
            if (words.find(w) == words.end()) {
                words.insert({w, 1});
            } else {
                words[w]++;
            }
        }
        next++;
    }

    int top = 0;
    for (auto iw = words.begin(); iw != words.end(); iw++) {
        top = max(top, iw->second);
    }

    vector<string> mostFrequently;
    for (auto iw = words.begin(); iw != words.end(); iw++) {
        if (iw->second == top) {
            mostFrequently.push_back(iw->first);
        }
    }

    sort(mostFrequently.begin(), mostFrequently.end());

    return mostFrequently;
}

vector<string> Amazon::reorderLines(int logFileSize, vector<string> logLines) {
    vector<string> words, integers;

    cout << endl;
    for (auto line : logLines) {
        regex r("([a-zA-Z]+[0-9]+)( [0-9]+)(.*)");
        if (regex_match(line, r)) {
            integers.push_back(line);
        } else {
            words.push_back(line);
        }
    }

    sort(words.begin(), words.end(), [](string& s1, string& s2) {
        regex r("^([A-Za-z]+)([0-9]+) (.+)");
        smatch m1, m2;
        bool b1 = regex_search(s1, m1, r);
        bool b2 = regex_search(s2, m2, r);
        if (b1 && b2) {
            if (m1[3] == m2[3]) {
                auto identifier_str1 = m1[1];
                auto identifier_str2 = m2[1];
                auto i1 = stoi(m1[2]);
                auto i2 = stoi(m2[2]);
                if (i1 == i2) {
                    return identifier_str1 < identifier_str2;
                }
                return i1 < i2;
            } else {
                return m1[3] < m2[3];
            }
        } else {
            return false;
        }
    });

//    sort(integers.begin(), integers.end(), [](string& s1, string& s2) {
//        regex r("^([A-Za-z]+)([0-9]+)");
//        smatch m1, m2;
//        bool b1 = regex_search(s1, m1, r);
//        bool b2 = regex_search(s2, m2, r);
//        if (b1 && b2) {
//            auto identifier_str1 = m1[1];
//            auto identifier_str2 = m2[1];
//            auto i1 = stoi(m1[2]);
//            auto i2 = stoi(m2[2]);
//            cout << "i1 = " << i1 << "; i2 = " << i2 << endl;
//            if (i1 == i2) {
//                return identifier_str1 < identifier_str2;
//            }
//            return i1 < i2;
//        } else {
//            return false;
//        }
//    });

    words.insert(words.end(), integers.begin(), integers.end());
    return words;
}
